Lib/ntpath.py

```diff
@@ -554,36 +554,48 @@ def normpath(path):
         return prefix + sep.join(comps)
 
 
-def _abspath_fallback(path):
-    """Return the absolute version of a path as a fallback function in case
-    `nt._getfullpathname` is not available or raises OSError. See bpo-31047 for
-    more.
-
-    """
-
-    path = os.fspath(path)
-    if not isabs(path):
-        if isinstance(path, bytes):
-            cwd = os.getcwdb()
-        else:
-            cwd = os.getcwd()
-        path = join(cwd, path)
-    return normpath(path)
-
 # Return an absolute path.
 try:
     from nt import _getfullpathname
 
 except ImportError: # not running on Windows - mock up something sensible
-    abspath = _abspath_fallback
+    def abspath(path):
+        """Return the absolute version of a path."""
+        path = os.fspath(path)
+        if not isabs(path):
+            if isinstance(path, bytes):
+                cwd = os.getcwdb()
+            else:
+                cwd = os.getcwd()
+            path = join(cwd, path)
+        return normpath(path)
 
 else:  # use native Windows method on Windows
     def abspath(path):
         """Return the absolute version of a path."""
         try:
             return _getfullpathname(normpath(path))
         except (OSError, ValueError):
-            return _abspath_fallback(path)
+            # See gh-75230, handle outside for cleaner traceback
+            pass
+        path = os.fspath(path)
+        if not isabs(path):
+            if isinstance(path, bytes):
+                sep = b'\\'
+                cwd = os.getcwdb()
+            else:
+                sep = '\\'
+                cwd = os.getcwd()
+            drive, root, path = splitroot(path)
+            if drive and drive != splitroot(cwd)[0]:
+                try:
+                    path = join(_getfullpathname(drive), path)
+                except (OSError, ValueError):
+                    # Invalid drive \x00: on Windows; assume root directory
+                    path = drive + sep + path
+            else:
+                path = join(cwd, root + path)
```

Comment:
If there's a `drive`, it must be a drive-relative path. If there's a `root`, it must be a rooted-relative path. There's no need to get the full working directory in these two cases. Calling `getcwd()` can be delayed until the branch that resolves a simple-relative path.

```suggestion
            if is_bytes := isinstance(path, bytes):
                sep = b'\\'
            else:
                sep = '\\'
            drive, root, path = splitroot(path)
            # Either drive or root can be nonempty, but not both.
            if drive or root:
                try:
                    path = join(_getfullpathname(drive + root), path)
                except (OSError, ValueError):
                    # Drive "\0:" cannot exist; use the root directory.
                    path = drive + sep + path
            else:
                cwd = os.getcwdb() if is_bytes else os.getcwd()
                path = join(cwd, path)
```

Reply:
Very clever, I've removed the walrus operator though.